core/src/main/java/hudson/lifecycle/Lifecycle.java

```diff
@@ -136,6 +139,19 @@ public void verifyRestartable() throws RestartNotSupportedException {
         return INSTANCE;
     }
 
+    /**
+     * Sets the singleton instance.
+     * Appropriate for implementations defined in plugins,
+     * since {@link #get} may be called before plugins are initialized,
+     * and so it is not safe to pass a plugin-defined class to the system property {@code hudson.lifecycle}
+     * despite the use of {@link PluginManager#uberClassLoader}.
```

Comment:
this comment is contradictory to me.

if this method is appropriate for plugins, but itself may be called before plugins are initialized, how can a plugin ensure that they can call `set` before `get` as any consumers of `get` can rightly assume that the lifecycle is fixed.
the answer is they can not - so it is not really "safe" to be called in a plugin.

once `get` is called - to all purposes `INSTANCE` should be final and never change.
You could probably demonstrate this by throwing an exception if `INSTANCE` is non null (which means get has been called)

so whilst setting the property is not safe (as you can not obtain load the correct lifecycle, you can not wait until plugins are loaded before you set the lifecycle either.

Reply:
> how can a plugin ensure that they can call `set` before `get`

It cannot, but it can call `set` in an `@Initializer` or otherwise during startup.

> any consumers of `get` can rightly assume that the lifecycle is fixed

Why would a consumer make that assumption? I did a search of @jenkinsci sources and found no example of a consumer doing anything with the return value other than immediately calling some method on it.

Reply:
> It cannot, but it can call set in an @Initializer or otherwise during startup.

which means if startup fails the lifecycle will not be correctly replaced leading to "the wrong" lifecyle handling the issue?

> Why would a consumer make that assumption

Because it was a single initialized only once in code.

> I did a search of https://github.com/jenkinsci sources and found no example of a consumer doing anything with the return value other than immediately calling some method on it.

an example (but a bad one) is https://github.com/jenkinsci/jenkins/blob/7d2b484b88e1700ebc62b6bca8fdd44d8ca56376/core/src/main/java/hudson/lifecycle/WindowsInstallerLink.java#L267-L288 that changes behaviour once and once only for example at startup as it has assumed that the Lifecycle will not change.

I hope to rip that example out in #8523 so this becomes a moot point.

I think this just needs some clearer documentation that this is super advanced and you should not normally do this - and it you do that you will not be called back during early initialisation like a regular `Lifecycle` and that even changing it could cause other plugin code that is obtaining the lifecycle (say to check if you can restart once) may not observe any changes in the lifecycle. (or documewnt in the get that the lifecycle *can* change at runtime)
esp the class level comment is now incorrect about [how the Lifecycle is found](https://github.com/jenkinsci/jenkins/blob/352bc190e518301421970399cac57ec199daaf66/core/src/main/java/hudson/lifecycle/Lifecycle.java#L53C1-L54).

Reply:
(for the avoidance of doubt my comments are non blocking)